packages/firestore/src/remote/bloom_filter.ts

```diff
@@ -0,0 +1,107 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { Md5, Integer } from '@firebase/webchannel-wrapper';
+
+import { newTextEncoder } from '../platform/serializer';
+import { debugAssert } from '../util/assert';
+
+const MAX_64_BIT_UNSIGNED_INTEGER = Integer.fromNumber(Math.pow(2, 64));
+export class BloomFilter {
+  private readonly size: Integer;
+
+  constructor(
+    private readonly bitmap: Uint8Array,
+    padding: number,
+    private readonly hashCount: number
+  ) {
+    const numBits = bitmap.length * 8 - padding;
+    this.size = Integer.fromNumber(numBits);
+
+    debugAssert(padding >= 0 && padding < 8, `Invalid padding: ${padding}`);
+    // Empty bloom filter should have 0 padding and 0 hash count.
+    if (bitmap.length === 0) {
+      debugAssert(
+        padding === 0 && hashCount === 0,
+        'A valid empty bloom filter will have all three fields be empty.'
```

Comment:
In the backend Bloom Filter implementation, an empty bloom filter should have all fields 0. cl488276956
```
  // A valid empty bloom filter will have all three fields be empty.
  if (hash_count == 0 && bitmap.empty() && num_paddings == 0) {
    return EmptyBloomFilter();
  }
```

Reply:
I've asked the backend team to consider relaxing this restriction. Even if they don't agree, I don't see a good reason to forbid hash functions on an empty bloom filter. But let's see what they say.